### DISCUSSION TITLE
Add SQLite//mariaDB support

### DISCUSSION THREAD

**Discussion by spacewalkingninja** [medusajs/medusa#4058] [Feature Requests]
Please add MySql//MariaDB support.
PostgreSQL is MAKING US USE 20 MB OF RAM FOR EVERY CONNECTION WE NEED TO MAKE!
I want to use this as a replacement for a prestashop/thirtybees store. 
This is unfair and very resource wasteful, not ecological at all.
All this because you want transactions and things that exist in postgresql.
Just insert and update things and stuff works like back when we were script kiddies that just used PHP and some CMS downloaded from the net.

Why am I suck a pain in you *sses:

> PostgreSQL starts a new process with its own memory allocation for each connection it establishes, so it requires a lot of memory on systems with a high number of client connections. MySQL uses a single process and maintains one thread (or path of execution) per connection, which results in less memory pressure.

**When to not use PostgreSQL**
1. When speed is imperative

What else do I need to explain, we are making online stores, we need them to be snappy fast, and run on 758mb of ram or whatebver the cheapest free VM we find on the net. 

When I started using MedusaJS I had a very careful consideration as to WHICH exxcatly DB should I use, SQLite or PostGres and after informing myself, I made the conscious decision to USE SQLITE AS IT IS BETTER FOR THE APPLICATION IN MIND. 
And now you want to do transactions? Just do 100 queries instead of 1 transaction, I think it is better for the user as it is faster.

**Comment by pevey** [comment]
My two cents is that I am glad support is being dropped for sqllite.  If sqllite support is taking any developer focus at all, I believe it is not worth it.

I wouldn't mind having support for mariadb.  I slightly prefer mariadb over postgres, but it's not a huge deal for me.

I think the question of what db to support is very much tied to the use of typeorm.  It doesn't seem like that package is still maintained.  It seems like the chances of our pull request on that project actually being merged are very slim, which means in addition to medusa, the dev team is now supporting a typeorm fork.

Has the core team considered whether it would be worth it at this point to move to something else, perhaps Prisma or Drizzle?  This could also potentially open the door to users choosing mariadb instead of postgres in a way that is abstracted by prisma and does not require additional development work.

Edit:  Also, mariadb absolutely supports transactions, and I think it is very important for any non-trivial database-driven app.  Anything that deals with money, including ecommerce, qualifies as non-trivial.

**Reply by pevey** [reply]
>  It doesn't seem like that package is still maintained. It seems like the chances of our pull request on that project actually being merged are very slim, which means in addition to medusa, the dev team is now supporting a typeorm fork.

I have to take back what I wrote since TypeORM just accepted the PR.

**Reply by olivermrbl** [reply]
> Has the core team considered whether it would be worth it at this point to move to something else, perhaps Prisma or Drizzle?

Glad that you bring this up. I've been meaning to write a couple of paragraphs on some of our recent discoveries as well as decisions related to ORMs. 

There are multiple things to touch upon here.

It has long been obvious to us, that Typeorm is not being maintained very actively - at least not to the degree that we see fit for our project. We’ve identified and reported several fixes and improvements in the past months that we’d like to see resolved, and we are yet to receive a response on some of those. 

Being an OS-library maintainer myself, I am well aware that the amount of issues and PRs can be overwhelming at times, so I don’t expect the guys to jump on this right away. Though, some of our reports were around critical issues that could be a decisive factor for whether companies would choose to work with Typeorm or not because they simply don’t work as expected (ironically, this is what ultimately was the case for us). So I would expect the issues to at least be addressed/commented on, so we know that eyes are still on the project. 

Upgrading to their latest major version was the drop that made the glass spill over. It spawned a range of unexpected issues - of those, some were around core functionality that most would deem as minimum requirements for an ORM. 

So to summarise, we don’t find Typeorm to be maintained actively enough and it contains too many issues for it to be a viable solution for Medusa. It has therefore been decided that we will switch to a different ORM, and we will do it now.

Before highlighting what ORM we have chosen, I would like to take a brief moment to explain what we will do differently when we switch to the new ORM.

In our current implementation, Typeorm tightly couples the service layer with the data layer. Replacing Typeorm requires a tremendous amount of work - to the extent of being almost impossible - and this somewhat contradicts the vision of what we want Medusa to be. We want to equip developers with highly customizable and extensible tooling that allows them to build any type of commerce application using (almost) any type of stack. The ORM is an element of the stack, so obviously, we think this should also be customizable and replaceable. 

As we replace Typeorm, we will create a generic data access layer, defining a set of methods to query data from your data store. This layer separates the business logic from the ORM and serves as a contract for what you will – as a minimum – have to implement if you wish to switch out the default ORM or connect to an entirely different type of storage provider. This gives developers the flexibility to choose the ORM that best fit their needs.

For the default ORM in Medusa, we've landed on [MikroORM](https://mikro-orm.io/). Our reasoning is simple. The API is very similar to Typeorm so the transition will be fast and straightforward. The bundle size is significantly smaller than Typeorm, which will become crucial as we move toward support for modern infrastructure solutions like serverless and edge runtime. And it comes with all of the features and concepts we need out of the box.

Concerning SQLite + MariaDB support, this will not be a priority to add to our core project now. Though, as we move to MikroORM and the new data access layer, you will be able to add this yourself by replacing or tweaking the underlying ORM and data store.

**Reply by albertobissacco** [reply]
When this ORM migration will be implemented?

**Reply by olivermrbl** [reply]
We are amid migrating away from Typeorm to MikroORM. The migration is done as part of refactoring each domain of our core into a separate, (optionally) standalone module - as we've done with `@medusajs/product`. 

Hence, this will be a long process, tackled incrementally over the next 6-12 months. As we get closer to completing the migration, we will share a guide on how to refactor existing Typeorm customizations and related functionality to the new ORM.

**Comment by spacewalkingninja** [comment]
Now I will have to consider how to put PHP + MedusaJS on my server so I can use adminer again. Another big problem of medusaJS. The beautiful simplicity of the Apache/ PHP / MySQL stack, is what we lost with dirty nodes and stuffs

**Comment by SGFGOV** [comment]
Perhaps this is a total newbie take on this, but I think a lot of users
will be affected by the switch, which I am sure you already know. Why not
just release your own version of the typeorm packages with the fixes that
have been proposed? We can use that for all Medusa. Already we have a major
change which is the extender depreciation, added to that this, it will be
perhaps a couple of months at least before all these changes are made and
tested. Significantly adding to costs and time. Won't you have to redo the
entire tooling including migrations, database loading etc.



On Tue, 16 May 2023 at 14:14, Oliver Windall Juhl ***@***.***>
wrote:

> Has the core team considered whether it would be worth it at this point to
> move to something else, perhaps Prisma or Drizzle?
>
> Glad that you bring this up. I've been meaning to write a couple of
> paragraphs on some of our recent discoveries as well as decisions related
> to ORMs.
>
> There are multiple things to touch upon here.
>
> It has long been obvious to us, that Typeorm is not being maintained very
> actively - at least not to the degree that we see fit for our project.
> We’ve identified and reported several fixes and improvements in the past
> months that we’d like to see resolved, and we are yet to receive a response
> on some of those.
>
> Being an OS-library maintainer myself, I am well aware that the amount of
> issues and PRs can be overwhelming at times, so I don’t expect the guys to
> jump on this right away. Though, some of our reports were around critical
> issues that could be a decisive factor for whether companies would choose
> to work with Typeorm or not because they simply don’t work as expected
> (ironically, this is what ultimately was the case for us). So I would
> expect the issues to at least be addressed/commented on, so we know that
> eyes are still on the project.
>
> Upgrading to their latest major version was the drop that made the glass
> spill over. It spawned a range of unexpected issues - of those, some were
> around core functionality that most would deem as minimum requirements for
> an ORM.
>
> So to summarise, we don’t find Typeorm to be maintained actively enough
> and it contains too many issues for it to be a viable solution for Medusa.
> It has therefore been decided that we will switch to a different ORM, and
> we will do it now.
>
> Before highlighting what ORM we have chosen, I would like to take a brief
> moment to explain what we will do differently when we switch to the new ORM.
>
> In our current implementation, Typeorm tightly couples the service layer
> with the data layer. Replacing Typeorm requires a tremendous amount of work
> - to the extent of being almost impossible - and this somewhat contradicts
> the vision of what we want Medusa to be. We want to equip developers with
> highly customizable and extensible tooling that allows them to build any
> type of commerce application using (almost) any type of stack. The ORM is
> an element of the stack, so obviously, we think this should also be
> customizable and replaceable.
>
> As we replace Typeorm, we will create a generic data access layer,
> defining a set of methods to query data from your data store. This layer
> separates the business logic from the ORM and serves as a contract for what
> you will – as a minimum – have to implement if you wish to switch out the
> default ORM or connect to an entirely different type of storage provider.
> This gives developers the flexibility to choose the ORM that best fit their
> needs.
>
> For the default ORM in Medusa, we've landed on MikroORM
> <https://mikro-orm.io/>. Our reasoning is simple. The API is very similar
> to Typeorm so the transition will be fast and straightforward. The bundle
> size is significantly smaller than Typeorm, which will become crucial as we
> move toward support for modern infrastructure solutions like serverless and
> edge runtime. And it comes with all of the features and concepts we need
> out of the box.
>
> Concerning SQLite + MariaDB support, this will not be a priority to add to
> our core project now. Though, as we move to MikroORM and the new data
> access layer, you will be able to add this yourself by replacing or
> tweaking the underlying ORM and data store.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/medusajs/medusa/discussions/4058#discussioncomment-5913826>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXEQJHAOKWDE4BQGGTBYUD3XGM463ANCNFSM6AAAAAAX3SX4BY>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

-- 

This message (including any attachments) may contain confidential, 
proprietary, privileged and/or private
information. The information is 
intended to be for the use of the individual or entity designated above. If
you are not the intended recipient of this message, please notify the 
sender immediately, and delete the
message and any attachments. Any 
disclosure, reproduction, distribution or other use of this message or
any 
attachments by an individual or entity other than the intended recipient is 
prohibited.

**Reply by pevey** [reply]
I've been glancing over the MikroORM docs, and it is really quite similar to TypeORM in its syntax.  I'm sure it will be a lot of work for the team, but maybe not quite as big a change to most users as you might think.

I also think being constrained by all the downsides of TypeORM has been a tremendous amount of work on its own.  Dealing with issues that are, at root, TypeORM issues, is a major distraction.  TypeORM is really holding Medusa back, IMO.  Yes, it will be a bit painful, but much better for the project in the long run to rip the bandaid off.  The sooner the better.
